### PR TITLE
upgrade to 0.1.4

### PR DESCRIPTION
**Pull request by ausgerechnet**
- AMs
  - bugfix in MI
  - signed version of log-likelihood
  - add log-ratio
- test data
  - add UCS gold
- naming conventions
  - follow Evert (2004: 36) *or* contingency table
- usability:
  - calculate observed and expected frequencies on the fly for AMs
  - add inplace operations
- extend README

Fixes https://github.com/fau-klue/pandas-association-measures/issues/12

### CODE REVIEW

**Comment by ausgerechnet** [comment]
@martialblog if you're happy please merge and deploy to PyPI

File: tests/test_measures_gold.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+from pandas import read_csv
+
+
+import association_measures.frequencies as fq
+import association_measures.measures as am
+
+
+def test_frequencies():
+
+    df = read_csv("tests/ucs-gold.ds.gz", comment='#', index_col=0,
```
Comment (martialblog) [inline tests/test_measures_gold.py:10]: I'm not a big fan of binary data in a repository, it'll become larger and larger when this changes. I'd prefer to host this somewhere else and download it, or have a smaller not-binary dataset

File: tests/test_readme.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import association_measures.frequencies as fq
+import association_measures.measures as am
+from pandas import read_csv
+
+
+def test_input():
+    df = read_csv("tests/ucs-gold.ds.gz", comment='#', index_col=0,
```
Comment (martialblog) [inline tests/test_readme.py:7]: Again, binary data

File: association_measures/frequencies.py [hunk 3]
```diff
@@ -1,54 +1,111 @@
...
+    naming conventions (cf. Evert 2004: 36)
+    f = O11                     # co-occurrence freq. of token and node
+    f1 = R1 = O11 + O12         # number of tokens in W(node)
+    f2 = C1 = O11 + O21         # marginal freq. of token
+    N                           # size of corpus
 
-    :param pandas.DataFrame df: Pandas Dataframe containing O11, f1, f2 and N
-    :return: Tuple of pandas.Series containing observed frequencies for tokens
+    :param pandas.DataFrame df: Pandas Dataframe containing O11 or f, f1, f2 and N
+    :return: tuple of pandas.Series (O11, O12, O21, O22)
```
Comment (martialblog) [inline association_measures/frequencies.py:19]: Add Docstring for new ValueErrors

**Comment by martialblog** [comment]
Hi, I left some reviews.

Also fixed the unittest, looks good but there are some Lint errors.

Also also, I don't think the inplace option is a good idea, as it probably will be deprecated: https://github.com/pandas-dev/pandas/issues/16529

Cheerio, 
Markus

**Comment by ausgerechnet** [comment]
Hi,

I agree on the inplace operation. I added this because your original implementation had some – but not consistent – inplace operations. For improved usability, we should thus return either the tuple of series as a dataframe (which can thus be joined with df.join(result) *or* the complete new df (with all the old columns). what do you think?

Where can I see addtional comments?

P

File: association_measures/measures.py [hunk 4]
```diff
@@ -8,9 +8,10 @@
 import pandas as pd
 import numpy as np
 from .binomial import choose as binomial
+from .frequencies import expected_frequencies
 
 
-choose = np.vectorize(binomial) # pylint: disable=invalid-name
+choose = np.vectorize(binomial)  # pylint: disable=invalid-name
```
[the remark this answers is not in the document]
Comment (ausgerechnet) [reply association_measures/measures.py:14]: [pep 8 guidelines](https://www.python.org/dev/peps/pep-0008/#inline-comments)

[the remark this answers is not in the document]
Comment (martialblog) [reply association_measures/measures.py:14]: :+1: Learn something new every day

File: tests/test_measures_gold.py (hunk 1 above)
Comment (ausgerechnet) [reply tests/test_measures_gold.py:10]: well the gold-data won't change and I doubt that Stefan will implement any new measures in UCS. but I see your point – I'll take a sample of all rows and check in uncompressed files.

File: setup.py [hunk 5]
```diff
@@ -47,11 +47,11 @@
 except FileNotFoundError:
     long_description = DESCRIPTION
 
-# Load the package's __version__.py module as a dictionary.
+# Load the package's version.py module as a dictionary.
```
[the remark this answers is not in the document]
Comment (ausgerechnet) [reply setup.py:50]: ocd

File: tests/test_measures.py [hunk 6]
```diff
@@ -48,8 +48,10 @@ def test_dice(sample_dataframe):
 def test_dice_nan(invalid_dataframe):
     df = invalid_dataframe
 
-    df['am'] = am.dice(df)
-    assert np.isnan(df['am'][0])
```
[the remark this answers is not in the document]
Comment (ausgerechnet) [reply tests/test_measures.py:52]: no idea what I'm doing wrong here …

File: tests/test_measures.py [hunk 7]
```diff
@@ -74,8 +76,10 @@ def test_t_score(sample_dataframe):
 def test_t_score_nan(invalid_dataframe):
     df = invalid_dataframe
 
-    df['am'] = am.t_score(df)
-    assert np.isnan(df['am'][0])
+    try:
```
[the remark this answers is not in the document]
Comment (ausgerechnet) [reply tests/test_measures.py:139]: see above

File: tests/test_measures.py [hunk 8]
```diff
@@ -96,12 +100,16 @@ def test_z_score(sample_dataframe):
...
+
 @pytest.mark.nan
 def test_z_score_nan(invalid_dataframe):
     df = invalid_dataframe
 
-    df['am'] = am.z_score(df)
-    assert np.isnan(df['am'][0])
+    try:
+        df['am'] = am.z_score(df)
+    except ValueError:
```
[the remark this answers is not in the document]
Comment (ausgerechnet) [reply tests/test_measures.py:171]: see above

File: tests/test_measures.py [hunk 9]
```diff
@@ -174,4 +186,4 @@ def test_with_random_data(random_dataframe):
     assert 'dice' in df.columns
     assert 'mutual_information' in df.columns
     assert 'log_likelihood' in df.columns
-    assert 'hypergeometric_likelihood' in df.columns
+    # assert 'hypergeometric_likelihood' in df.columns
```
[the remark this answers is not in the document]
Comment (ausgerechnet) [reply tests/test_measures.py]: current implementation is super-slow, original implementation is buggy

**Comment by ausgerechnet** [comment]
- I removed inplace operations.
-

**Comment by ausgerechnet** [comment]
I think we're good here. Can you have another look and merge it or request further changes?

File: association_measures/measures.py [hunk 10]
```diff
@@ -5,149 +5,202 @@
...
 
 
-import pandas as pd
 import numpy as np
-from .binomial import choose as binomial
+from .frequencies import expected_frequencies, observed_frequencies
 
 
-choose = np.vectorize(binomial) # pylint: disable=invalid-name
+def choose(n, k):
```
Comment (martialblog) [inline association_measures/measures.py:12]: Can we just fix the choose function as described and not have a second implementation?

See https://github.com/fau-klue/pandas-association-measures/issues/12

File: association_measures/measures.py [hunk 11]
```diff
@@ -5,149 +5,202 @@
...
-
-    res = (2 * df['O11']) / (df['f1'] + df['f2'])
+    am = (2 * freq['O11']) / (2 * freq['O11'] + freq['O12'] + freq['O21'])
 
-    return pd.Series(data=res)
+    return am
 
 
-def log_likelihood(df):
+def log_likelihood(freq, signed=True):
```
Comment (martialblog) [inline association_measures/measures.py:105]: I would like these functions to be as polymorphic as possible, can we maybe not have a parameter but just return an extra column with the signed valued?

File: tests/test_binomial.py [hunk 12]
```diff
@@ -1,8 +1,10 @@
 import pytest
 
 from association_measures import binomial as bi
+from association_measures.measures import choose
 
 
+@pytest.mark.skip
```
Comment (martialblog) [inline tests/test_binomial.py:7]: Again, See https://github.com/fau-klue/pandas-association-measures/issues/12

File: tests/test_readme.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
+import association_measures.frequencies as fq
+import association_measures.measures as am
+from pandas import read_csv
+
+
+def test_input():
```
Comment (martialblog) [inline tests/test_readme.py:6]: These tests don't assert anything?

File: tests/test_measures.py [hunk 14]
```diff
@@ -1,177 +1,241 @@
...
+@pytest.mark.zero
 def test_hypergeometric_likelihood_with_zeros(zero_dataframe):
     df = zero_dataframe
-
-    df['am'] = am.hypergeometric_likelihood(df)
-    assert df['am'][0] == 1.0
-
-
-# Check if any warnings of errors are thrown. Might be an unstable test
-@pytest.mark.stability
```
Comment (martialblog) [inline tests/test_measures.py:167]: Why was this removed?

**Comment by martialblog** [comment]
Also the Python Lint is still failing.

File: tests/test_measures.py (hunk 14 above)
Comment (ausgerechnet) [reply tests/test_measures.py:167]: random data may lead to ValueErrors (if e.g. O11 > N) – this test is thus not possible anymore. The alternative would be to return NAN values for wrong rows and throw a warning instead of an error. Do you prefer this?

File: tests/test_readme.py (hunk 13 above)
Comment (ausgerechnet) [reply tests/test_readme.py:6]: this is my way of creating the output for the README. it doesn't test anything in addition to all the other tests. you can change it if you like!

File: tests/test_binomial.py (hunk 12 above)
Comment (ausgerechnet) [reply tests/test_binomial.py:7]: works like a charm with new type

File: association_measures/measures.py (hunk 11 above)
Comment (ausgerechnet) [reply association_measures/measures.py:105]: I don't really understand the issue of polymorphism here, but I don't really mind.

I just removed the unsigned version, nobody should use this one anyway …

File: association_measures/measures.py (hunk 10 above)
Comment (ausgerechnet) [reply association_measures/measures.py:12]: yes, we can!

and I did.

**Comment by ausgerechnet** [comment]
- changed some coding according to the pylint conventions
- excluded "too-many-branches" from pylint
- changed pylint naming conventions of variables so that variable names can follow the generally accepted terminology

File: tests/test_measures.py (hunk 14 above)
Comment (martialblog) [reply tests/test_measures.py:167]: Let me have a closer look at it
